### PR TITLE
[FEA] Add support for `cudf.Timestamp`

### LINKED ISSUE #12052: [FEA] cudf.Timestamp
**Is your feature request related to a problem? Please describe.**
rewriting code from pandas into cudf, using `import cudf as pd`

**Describe the solution you'd like**
`cudf.Timestamp` matching https://pandas.pydata.org/docs/reference/api/pandas.Timestamp.html

### LINKED ISSUE #16420: [BUG] Series.min returns `np.datetime64` not `pd.Timestamp` in `cudf.pandas`
**Describe the bug**
Calling `min` or `max` on a column with `pd.Timestamp` data causes the timestamp to be converted to `np.datetime64`.

**Steps/Code to reproduce bug**
```
# with cudf.pandas loaded
s = pd.Series([pd.Timestamp("2018-01-05")])
s.min()
numpy.datetime64('2018-01-05T00:00:00.000000000')
```
**Expected behavior**
```
# without cudf.pandas loaded
s = pd.Series([pd.Timestamp("2018-01-05")])
s.min()
Timestamp('2018-01-05 00:00:00')
```
**Environment overview (please complete the following information)**
- bare-metal, from source

**Environment details**
Please run and paste the output of the `cudf/print_env.sh` script here, to gather any other relevant environment details

**Additional context**

### PR DESCRIPTION
**Pull request by Matt711**
## Description

closes #12052 and closes #16420

## Checklist
- [ ] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [ ] New or existing tests cover these changes.
- [ ] The documentation is up to date with these changes.

### CODE REVIEW

File: python/cudf/cudf/core/timestamp.py [hunk 1]
```diff
@@ -0,0 +1,389 @@
...
+    def astimezone(self, tz=None):
+        raise NotImplementedError(
+            "The method 'astimezone' is not implemented."
+        )
+
+    def ceil(self, freq=None):
+        raise NotImplementedError("The method 'ceil' is not implemented.")
+
+    @classmethod
+    def combine(cls, date, time):
```
Comment (Matt711) [inline python/cudf/cudf/core/timestamp.py:247]: TODO: Type the arguments

File: python/cudf/cudf/core/timestamp.py [hunk 2]
```diff
@@ -0,0 +1,389 @@
...
+
+from datetime import datetime, timedelta
+
+import numpy as np
+import pandas as pd
+
+from cudf.core.scalar import Scalar
+
+
+class Timestamp(Scalar):
```
Comment (Matt711) [inline python/cudf/cudf/core/timestamp.py:13]: I wonder if would be better to first get the implementations of `cudf.Timestamp` and `cudf.Timedelta` in and track their performance implications before replacing them throughout cudf (eg. [here](https://github.com/rapidsai/cudf/blob/ac42bc870a65d807784cae63e25b9e9ca788eb23/python/cudf/cudf/core/column/datetime.py#L574))

**Review by Matt711 — COMMENTED** [review]
I think the bulk of the changes are there. It still needs some tests.

cc. @mroeschke

File: python/cudf/cudf/core/timestamp.py [hunk 3]
```diff
@@ -0,0 +1,389 @@
...
+        self,
+        *args,
+        **kwargs,
+    ):
+        pd_ts_kwargs = {k: v for k, v in kwargs.items() if k != "dtype"}
+        ts = pd.Timestamp(*args, **pd_ts_kwargs)
+        super().__init__(ts)
+
+    @property
+    def value(self) -> np.datetime64:
```
Comment (mroeschke) [inline python/cudf/cudf/core/timestamp.py:24]: Should the host value just be a `pandas.Timestamp`?

File: python/cudf/cudf/core/timestamp.py [hunk 4]
```diff
@@ -0,0 +1,389 @@
...
+    @property
+    def microsecond(self) -> int:
+        return pd.Timestamp(super().value).microsecond
+
+    @property
+    def nanosecond(self) -> int:
+        return pd.Timestamp(super().value).nanosecond
+
+    def __repr__(self):
+        return pd.Timestamp(self.value).__repr__()
```
Comment (Matt711) [inline python/cudf/cudf/core/timestamp.py:60]: TODO: `super().value` or add a new property `_scalar_value` which returns `super().value`

File: python/cudf/cudf/core/timestamp.py [hunk 5]
```diff
@@ -0,0 +1,389 @@
...
+    @classmethod
+    def from_pandas(cls, obj: pd.Timestamp):
+        return cls(obj)
+
+    @classmethod
+    def from_scalar(cls, obj: Scalar):
+        return cls(obj.value)
+
+    def _to_scalar(self):
+        return Scalar(self.value)
```
Comment (Matt711) [inline python/cudf/cudf/core/timestamp.py:78]: TODO: super().value

File: python/cudf/cudf/core/timestamp.py [hunk 6]
```diff
@@ -0,0 +1,389 @@
...
+
+    def __add__(self, other: timedelta | np.timedelta64):
+        return self.from_scalar(self._to_scalar() + other)
+
+    def __radd__(self, other: timedelta):
+        return self + other
+
+    def __sub__(
```
Comment (Matt711) [inline python/cudf/cudf/core/timestamp.py:86]: TODO: super().value

File: python/cudf/cudf/core/timestamp.py (hunk 3 above)
Comment (Matt711) [reply python/cudf/cudf/core/timestamp.py:24]: So it looks like there's a `cudf.Scalar.value` which `cudf.Timestamp` inherits. It returns a `np.datetime64` object. But there's also `pd.Timestamp.value` which returns an integer.

I think your right that`cudf.Timestamp.value` should return `pd.Timestamp(super().value).value`. And I'll a separate property for access to the scalar value (np.datetime64).

File: python/cudf/cudf_pandas_tests/test_profiler.py [hunk 7]
```diff
@@ -47,10 +47,12 @@ def test_profiler():
         "Timedelta",
         "_Timestamp.__add__",
     }
+    # TODO: Swap _Timestamp.__add__ for Timestamp.__add__
+    # when cudf.Timedelta support is added
     for name, func in per_function_stats.items():
         assert (
             len(func["cpu"]) == 0
-            if "Time" not in name
+            if "Timedelta" not in name or "_Time" not in name
```
Comment (Matt711) [inline python/cudf/cudf_pandas_tests/test_profiler.py:55]: TODO: Look into test failure caused here

**Review by Matt711 — COMMENTED** [review]
Should be ready for review soon after these comments are addressed and I add some tests.

File: python/cudf/cudf/core/timestamp.py [hunk 8]
```diff
@@ -0,0 +1,389 @@
...
+
+    @classmethod
+    def from_pandas(cls, obj: pd.Timestamp):
+        return cls(obj)
+
+    @classmethod
+    def from_scalar(cls, obj: Scalar):
+        return cls(obj.value)
+
+    def _to_scalar(self):
```
Comment (Matt711) [inline python/cudf/cudf/core/timestamp.py:77]: Make it public like `from_scalar`
```suggestion
    def to_scalar(self):
```

File: python/cudf/cudf/core/timestamp.py [hunk 9]
```diff
@@ -0,0 +1,395 @@
...
+# TODO: Use cupy when it supports timestamps and timedeltas. See https://github.com/cupy/cupy/issues/2622
+# TODO: Replace with cudf.Timedelta. See https://github.com/rapidsai/cudf/issues/5882
+class Timestamp(Scalar):
+    def __init__(
+        self,
+        *args,
+        **kwargs,
+    ):
+        pd_ts_kwargs = {k: v for k, v in kwargs.items() if k != "dtype"}
+        ts = pd.Timestamp(*args, **pd_ts_kwargs)
```
Comment (galipremsagar) [inline python/cudf/cudf/core/timestamp.py:22]: Is there a reason you are constructing a pandas Timestamp here? Is it being done only for validation purposes?

**Comment by Matt711** [comment]
I'd like to have some more offline discussion about the implementations of `cudf.Timestamp` and `cudf.Timedelta` before I proceed. Closing the PR for now.

**Pull request closed without merging**